Add cache to Variables class.

e-e-e opened the pull request:
## What did you implement:

Partially Closes #3821

This is specifically to deal with SSM Rate Exceeded errors.
It is a complement to https://github.com/serverless/serverless/pull/4294

We are having issues with Serverless calling SSM over 50 times because around 10 variables are referenced in a number of places. This issue is currently stopping us from deploying.

## How did you implement it:

I added a cache object to Variables class. This is referenced before AWS SSM is requested.

## How can we verify it:

I cannot share our config as our ssm variables are secrets. But this change is a pretty simple. The tests verify that the cache is working.

## Todos:

- [x] Write tests
- [x] Write documentation - N/A
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

HyperBrain commented:
Hi @e-e-e . Thanks for the PR.

I'm not really sure if this is a valid fix for the rate exceeded (for SSM it sure is). But imo we need a global solution, as otherwise, we have to remove this code again as soon as a real solution is at hand.
The problem of ridiculous multi-resolving of single variables does affect ALL variables, not only SSM. E.g. CF variables also lead to the issue, because they do external AWS API calls too.

@horike37 @RafalWilinski What's your opinion on this?

e-e-e commented:
@HyperBrain Yeah totally understand. I would be interested in helping on that global solution too. Just wanted to solve the immediate issue for us which is halting our deployment.
Are there thoughts towards a global solution atm? Are you thinking this should be at the provider level rather than Variable, or...?

ps. I really love serverless - its a great project and has made my life so much easier. So thanks for the sharing the work.

RafalWilinski commented:
Thanks for contribution @e-e-e, we really appreciate it! 💯 

I feel like this solution might not work ideally or as we would like to. I think If you add a `console.log` statement to see when these calls are made, you'll see that they are made all at once and all cache requests in this case will be cache misses because all of the requests are made before first requests completes.

When it comes to test, it works because it doesn't apply any latency. Response is immediate and that's why second call gets value from cache instead of requesting it from provider. 

I also agree with @HyperBrain, we should aim for global solution. Adding something like a queue on Bluebird layer (I think you mentioned something like that in past Frank) might be a quick win. (Or not so quick actually)

e-e-e commented:
@RafalWilinski Thanks for the quick reply. I updated the test to show that the cache is indeed being hit.  I am running my fork locally and it does indeed solve the issue for us. First thing I did was add a console log to check if it was hitting the cache or refetching the results.

Promise queue makes sense, but I wonder if there are actually two separate issues here. First the unnecessary refetching of variables that we can be certain have not changed (ssm, cf, etc), and then the rate limiting generally. Because even if we cache the results of some api calls, the rate limit can still be hit. Or alternatively if you add a promise queue without caching then the deployment time will be longer than needed as it will be waiting to refetch the same vars each time.

Is it possible to address the caching issue first, and then open a new issue/PR for addressing the queuing. I wrote a small native es6 library to help with concurrent promise queueing which might be useful - https://github.com/e-e-e/a-promise-queue - it lets you optionally select promise flavour. Although also as your using bluebird already maybe https://github.com/zackiles/bluebird-queue is the better choice.

To address caching - it would be trivial to set up similiar caches like the one I have implemented for each of the calls we know will always return the same response. Perhaps the functions themselves could be wrapped in [_.memoize](https://lodash.com/docs/4.17.4#memoize), if you're adverse to implementing adhoc caches.

Happy to help resolve this issue, as its really halting our development at the moment.

Also for some reason after adding a delay to my test, one of the travis instances is failing now. I dont think its related to my change if you look at the failure points. Have you seen these failures before? Or have I done something silly.

horike37 commented:
Thank you for moving forward this issue @e-e-e :+1: 

>Is it possible to address the caching issue first, and then open a new issue/PR for addressing the >queuing.

Yes. I think it would be good to separate issue/PR for different layers. It would make easier to review and test it for us :+1: 

I also agree with everyone here, then just want to confirm my understanding.
You just stated two topics which are caching and queueing solutions in above comment.
I could understand the value of caching one, then is queueing one to make the second request(after caching) to access cached valuables properly?

>Also for some reason after adding a delay to my test, one of the travis instances is failing now. I dont think its related to my change if you look at the failure points. Have you seen these failures before? Or have I done something silly.

The failure didn't be caused by this PR. I retired the Travis instance and seeing to fix the test for now. The tests on Travis sometimes fail but it is out of the scope of this PR.

e-e-e commented:
@horike37 also thank you for the promptness and input.

You all understand the code base more intimately then I do, but my feeling would be to the caching could happen in the Variable class, while the promise queue / rate limiting could be implemented at the Provider level.

> I could understand the value of caching one, then is queueing one to make the second request(after caching) to access the cached variables properly?

Caching would stop unneeded requests, while the promise queue would ideally add rate limiting. e.g. if 60 unique SSM or CF variables are requested, caching does not help but promise queue would. The question for me is also around whether AWS rate limits per endpoint or per account. As that would impact how you choose to implement queuing - whether there is a queue per endpoint or a queue per provider.

RafalWilinski commented:
Sorry @e-e-e, my bad. Seems my analysis wasn't good/deep enough. If it resolves an issue for you then it's great.

HyperBrain commented:
@e-e-e Fully agree with the approach you proposed. The caching should be part of the general variable handling and needs to be part of the generic Variables class and everything that is specific to the provider (like queueing the requests to prevent rate limits in case you have lots of different variables that have to be resolved) belong to the provider plugin.
Both together will solve the problem as a whole.

Imo we can do these two in separate PRs (one for the generic caching) and one for the AWS request queuing. The AWS request queuing should be anchored in the request class, where the AWS SDK functions are called (this is central for the AWS plugin).

e-e-e commented:
@HyperBrain totally agree. I just updated this PR to address generic variable caching.

I applied it to all the variables, however it's also quite trivial now to exclude certain variable types from the cache. All that is needed is to return from the if statement in `getValueFromSource()` rather than assigning the returned value to the cache.

Let me know your thoughts?

e-e-e commented on lib/classes/Variables.js:
```diff
@@ -167,23 +168,33 @@ class Variables {
+    }
+    if (valueFromSource) {
+      this.cache[variableString] = valueFromSource;
+      return valueFromSource;
```
Awesome. I have just updated this - so that once the promise is resolved it replaces itself in the cache with the resolved value.

I had similar thoughts to you when initial writing this - that ideally you should not cache a promise. But by caching the promise, the promise itself takes care of the `waiters` array in your pseudo code. As while the promise is yet to be resolved, any new call for the variable gets the promise that will eventually become the resolved value. Also if there are multiple requests, the appended .then()’s are executed in order of attachment - so it the promise acts as its own queue. This ends up doing the same thing as your proposed implementation but with less complexity.

HyperBrain commented on lib/classes/Variables.js:
```diff
@@ -167,23 +168,37 @@ class Variables {
+        this.cache[variableString] = result
+        return result;
+      });
+      return valueFromSource;
```
We should return the whole promise here - exactly the one that is stored in the cache map, so that the first invoker gets the same as all subsequent ones. This will guarantee consistent behavior.
```js
const resolver = BbPromise.resolve(valueFromSource)
  .then((result) => {
     this.cache[variableString] = result
     return result;
  });
this.cache[variableString] = resolver;
return resolver;
```

e-e-e replied:
Ah good catch! Thanks for the review feedback - I will fix it and eslint now.
I am really appreciative of how responsive and considered you and the serverless team are.

HyperBrain commented:
@e-e-e Nice catch 👍 . It is now really lean and maintainable. I only had one comment about the return value for the first invoker.

e-e-e commented:
@HyperBrain This one is ready for your eyes again.

HyperBrain commented:
Tagged for next release - merging

e-e-e commented:
Thanks again!
